### DISCUSSION TITLE
Question on contract instances

### DISCUSSION THREAD

**Discussion by orthecreedence** [freenet/freenet-core#1027] [Q&A] [answered]
To my understanding, a contract is a WASM program (I'm assuming a hash of it). Does the key for this program contain metadata? In other words, can two instances of the same WASM program exist under different keys, but also (verifiably) running the exact same code?

Or can only one instance of a WASM program exist in any given network?

 Following this, if multiple instances of the same contract are supported, would it be possible for two contracts to interact with eachother only if they share the same code (or some previous version of some known code)?

Thanks!

**Comment by sanity** [comment, marked as answer]
> To my understanding, a contract is a WASM program (I'm assuming a hash of it). Does the key for this program contain metadata? In other words, can two instances of the same WASM program exist under different keys, but also (verifiably) running the exact same code?

Sorry for the slow reply. Contracts are identified on the network by:
```
id = hash(hash(wasm) + parameters)
```

Where the hash algorithm is Blake3.

The parameters are an arbitrary block of bytes that's typically used to configure the contract - for example it would often contain the contract owner's public key.

So you could have two contracts with the same wasm code but different parameters and these would have completely different identifiers, but if both code and parameters are the same then it would be the same contract.

> Following this, if multiple instances of the same contract are supported, would it be possible for two contracts to interact with eachother only if they share the same code (or some previous version of some known code)?

A contract can read the state of *any* other contract (including contracts with the same wasm code) using the "related contracts" mechanism (which is under-documented right now, but see [contract_interface.rs](https://github.com/freenet/freenet-stdlib/blob/main/rust/src/contract_interface.rs)).

**Reply by orthecreedence** [reply]
Ok, thank you. This is what I expected but wanted to make sure.
